### PR TITLE
Enforce names in map_partitions

### LINKED ISSUE #620: Series vs. DataFrame confusion
I'm applying a function that returns a scalar to a group (alldata is a Dask dataframe):

```
def func(gp):
     return gp['Ref'].iloc[0]

estimates = alldata.groupby('Id').apply(func)
estimates.columns = ['rowno', 'Expected']   # unlike Pandas, this doesn't change column names
estimates.to_csv('out.csv',  index=False, index_label=False)  # doesn't work because estimates is not a DataFrame.  Instead it seems to be a Series
```

I find that estimates is now a Series, and Series doesn't support writing out. How can I write out all the values in this Series to disk? There is no values() method. I tried to work around this by using head()
```
estimates = estimates.head(n=len(estimates))
pd.DataFrame({'Id': np.arange(1,len(estimates)+1), 'Expected': estimates}).to_csv(...)
```
but this doesn't work either because only the first partition is written out, not the full file.

### LINKED ISSUE #627: Partition boundaries not honoring index
I may be misunderstanding dask, but I think the results should be independent of where the data are partitioned.  That doesn't seem to happen.  You can observe that the results change depending on whether npartitions is 2 or 1.

```
#!/usr/bin/env python
#
import dask.dataframe as dd
import pandas as pd

npartitions = 2  # change this to 1 to get correct result

outfile = '/tmp/out.csv'
df = pd.DataFrame({'uid': [1,1,2,2,2],
                   'Count': [43.,42.,43.,21.,45.]})
alldata = dd.from_pandas(df, npartitions)
alldata.set_index('uid', inplace=True)
print alldata.head()

def myfunc(gp):
    return gp['Count'].sum()

estimates = alldata.groupby('uid').apply(myfunc, columns=['uid', 'total'])
estimates.to_csv(outfile, header=True)

# print out file that just got written
infile = open(outfile, 'r')
print infile.read()
infile.close()
```

### PR DESCRIPTION
**Pull request by mrocklin**
This ensures that the names of the partitions match the column names
given in ``map_partitions`` calls

Fixes #620
Fixes #627

### CODE REVIEW

File: dask/dataframe/core.py [hunk 1]
```diff
@@ -1595,17 +1593,31 @@ def map_partitions(func, columns, *args, **kwargs):
...
 
     dasks = [arg.dask for arg in args if isinstance(arg, _Frame)]
     return return_type(merge(dsk, *dasks), name, columns, args[0].divisions)
 
 
+def _rename(columns, df):
+    """ Rename columns in dataframe or series """
+    if isinstance(columns, Iterator):
+        columns = list(columns)
+    if columns is None:
```
Comment (sinhrks) [inline dask/dataframe/core.py]: Should use ``no_default`` to allow ``Series`` name reset to ``None``

Comment (mrocklin) [inline dask/dataframe/core.py]: Fixed

**Comment by mrocklin** [comment]
Merging soon if no comments.

**Comment by mrocklin** [comment]
Merging soon if no comments.

**Pull request merged**